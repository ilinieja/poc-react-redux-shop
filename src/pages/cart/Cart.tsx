import { useSelector } from "react-redux";
import { Button, ButtonVariants } from "shared/Button/Button";

import { ProductsListing } from "shared/ProductsListing/ProductsListing";
import { cartProductsSelectors, selectTotalPrice } from "store/selectors";

import styles from "./Cart.module.css";

export function Cart() {
  const productIds = useSelector(cartProductsSelectors.selectIds);
  const totalPrice = useSelector(selectTotalPrice);

  return (
    <div className={styles.container}>
      {productIds.length ? (
        <div className={styles.row}>
          <h3 className={styles.title}>
            Total price: <span className={styles.largeFont}>{totalPrice}</span>
          </h3>
          <Button
            variant={ButtonVariants.filled}
            onClick={() =>
              alert(
                "Thank you. Checkout is not implemented yet, so just take the letters for free."
              )
            }
          >
            Checkout
          </Button>
        </div>
      ) : (
        <div className={styles.row}>
          <h3 className={styles.title}>Nothing here yet...</h3>
        </div>
      )}
      <ProductsListing productIds={productIds} />
    </div>
  );
}
