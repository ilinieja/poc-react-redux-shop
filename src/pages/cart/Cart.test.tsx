import { renderWithProviders, TEST_IDS } from "testing/testing";
import { Cart } from "./Cart";
import { fireEvent, screen } from "@testing-library/react";

it("renders bought products", async () => {
  renderWithProviders(<Cart />);
  expect(
    screen
      .getAllByTestId(TEST_IDS.productLetterTitle)
      .map((el) => el.textContent)
  ).toMatchSnapshot();
});

it("renders correct quantities of bought products", async () => {
  renderWithProviders(<Cart />);
  expect(
    screen
      .getAllByTestId(TEST_IDS.productLetterBoughtQuantity)
      .map((el) => el.textContent)
  ).toMatchSnapshot();
});

it("renders correct total price", async () => {
  renderWithProviders(<Cart />);

  expect(screen.getByTestId(TEST_IDS.cartTotal).textContent).toMatchSnapshot();
});

it("removes product from listing when it's 1 and decremented", async () => {
  renderWithProviders(<Cart />);

  expect(
    screen
      .getAllByTestId(TEST_IDS.productLetterTitle)
      .map((el) => el.textContent)
  ).toMatchSnapshot();

  fireEvent.click(screen.getAllByTestId(TEST_IDS.productDecrementButton)[2]);

  expect(
    screen
      .getAllByTestId(TEST_IDS.productLetterTitle)
      .map((el) => el.textContent)
  ).toMatchSnapshot();
});

it("updates total price when product is decremented", async () => {
  renderWithProviders(<Cart />);

  expect(screen.getByTestId(TEST_IDS.cartTotal).textContent).toMatchSnapshot();

  fireEvent.click(screen.getAllByTestId(TEST_IDS.productDecrementButton)[0]);

  expect(screen.getByTestId(TEST_IDS.cartTotal).textContent).toMatchSnapshot();
});

it("updates total price when product is incremented", async () => {
  renderWithProviders(<Cart />);

  expect(screen.getByTestId(TEST_IDS.cartTotal).textContent).toMatchSnapshot();

  fireEvent.click(screen.getAllByTestId(TEST_IDS.productIncrementButton)[0]);

  expect(screen.getByTestId(TEST_IDS.cartTotal).textContent).toMatchSnapshot();
});

it("updates total price when special offer stops working", async () => {
  renderWithProviders(<Cart />);

  expect(screen.getByTestId(TEST_IDS.cartTotal).textContent).toMatchSnapshot();

  fireEvent.click(screen.getAllByTestId(TEST_IDS.productDecrementButton)[1]);

  expect(screen.getByTestId(TEST_IDS.cartTotal).textContent).toMatchSnapshot();
});
