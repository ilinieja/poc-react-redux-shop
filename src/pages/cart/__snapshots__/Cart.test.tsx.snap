// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removes product from listing when it's 1 and decremented 1`] = `
Array [
  "C",
  "A",
  "D",
  "F",
  "V",
]
`;

exports[`removes product from listing when it's 1 and decremented 2`] = `
Array [
  "C",
  "A",
  "F",
  "V",
]
`;

exports[`renders bought products 1`] = `
Array [
  "C",
  "A",
  "D",
  "F",
  "V",
]
`;

exports[`renders correct quantities of bought products 1`] = `
Array [
  "5",
  "3",
  "1",
  "2",
  "3",
]
`;

exports[`renders correct total price 1`] = `"395"`;

exports[`updates total price when product is decremented 1`] = `"395"`;

exports[`updates total price when product is decremented 2`] = `"375"`;

exports[`updates total price when product is incremented 1`] = `"395"`;

exports[`updates total price when product is incremented 2`] = `"415"`;

exports[`updates total price when special offer stops working 1`] = `"395"`;

exports[`updates total price when special offer stops working 2`] = `"365"`;
